src/Randomizer.SMZ3/Logic.cs

```diff
@@ -1,20 +1,140 @@
+        }
+
+        public bool CanAccessDeathMountainPortal(Progression items)
+        {
+            return (CanDestroyBombWalls(items) || items.SpeedBooster) && items.Super && items.Morph;
+        }
+
+        public bool CanAccessDarkWorldPortal(Progression items)
+        {
+            return items.CardMaridiaL1 && items.CardMaridiaL2 && CanUsePowerBombs(items) && items.Super && items.Gravity && items.SpeedBooster;
+        }
+
+        public bool CanAccessMiseryMirePortal(Progression items)
+        {
+            return (items.CardNorfairL2 || (items.SpeedBooster && items.Wave)) && items.Varia && items.Super && (items.Gravity && items.SpaceJump) && CanUsePowerBombs(items);
+        }
+
+        public bool CanIbj(Progression items)
+        {
+            return items.Morph && items.Bombs && World.Config.LogicConfig.InfiniteBombJump;
+        }
+
+        public bool CanFly(Progression items)
+        {
+            return items.SpaceJump || CanIbj(items);
+        }
+
+        public bool CanUsePowerBombs(Progression items)
+        {
+            return items.Morph && (!World.Config.LogicConfig.RequireTwoPowerBombs && items.PowerBomb) || items.TwoPowerBombs;
+        }
+
+        public bool CanPassBombPassages(Progression items)
+        {
+            return items.Morph && (items.Bombs || CanUsePowerBombs(items));
+        }
+
+        public bool CanDestroyBombWalls(Progression items)
+        {
+            return CanPassBombPassages(items) || CanSafelyUseScrewAttack(items);
+        }
+
+        public bool CanSpringBallJump(Progression items)
+        {
+            return items.Morph && items.SpringBall;
+        }
+
+        public bool CanHellRun(Progression items)
+        {
+            return items.Varia || HasEnergyReserves(items, 5);
+        }
+
+        public bool HasEnergyReserves(Progression items, int amount)
+        {
+            return (items.ETank + items.ReserveTank) >= amount;
+        }
+
+        public bool CanOpenRedDoors(Progression items)
+        {
+            return items.Missile || items.Super;
+        }
+
+        public bool CanAccessNorfairUpperPortal(Progression items)
+        {
+            return items.Flute || (CanLiftLight(items) && items.Lamp);
+        }
+
+        public bool CanAccessNorfairLowerPortal(Progression items)
+        {
+            return items.Flute && CanLiftHeavy(items);
+        }
+
+        public bool CanAccessMaridiaPortal(Progression items)
+        {
+            return items.MoonPearl && items.Flippers &&
+                    items.Gravity && items.Morph &&
+                    (World.CanAquire(items, Reward.Agahnim) || items.Hammer && CanLiftLight(items) || CanLiftHeavy(items));
+        }
+
+        public bool CanSafelyUseScrewAttack(Progression items)
+        {
+            return items.ScrewAttack && (!World.Config.LogicConfig.PreventScrewAttackSoftLock || items.Morph);
+        }
+
+        public bool CanFireRodLightDarkRooms(Progression items)
```

Comment:
Considering this check is almost always used in conjunction with ` || items.Lamp`, maybe just make that part of this check? Mind you, I'm not too familiar with the fire rod logic thing.

Reply:
Updated function name to CanPassFireRodDarkRooms and made this change